test/fika/compiler/type_checker_test.exs

```diff
@@ -152,6 +152,87 @@ defmodule Fika.Compiler.TypeCheckerTest do
     assert {:ok, :Int} = TypeChecker.infer(foo, env)
   end
 
+  test "infer return type when there is simple recursion" do
+    str = """
+    fn factorial(x: Int, acc: Int) : Int do
+      if x <= 1 do
+        acc
+      else
+        factorial(x - 1, acc * x)
+      end
+    end
+    """
+
+    {:ok, ast} = Parser.parse_module(str)
+    [factorial] = ast[:function_defs]
+
+    env = TypeChecker.init_env(ast)
+
+    assert {:ok, %T.Loop{type: :Int}} = TypeChecker.infer(factorial, env)
+  end
+
+  test "infer return type when there is intra-module recursion" do
+    str = """
+    fn foo(a: Int) : Int do
+      bar(a)
+    end
+
+    fn bar(a: Int) : Int do
+      if a <= 1 do
+        foobar()
+      else
+        baz(a - 1)
+      end
+    end
+
+    fn baz(a: Int) : Int do
+      if a == 1 do
+        1
+      else
+        foo(a - 1)
+      end
+    end
+
+    fn foobar : Nothing do
+      to_atom("Nothing")
+    end
+
+    ext to_atom(x: String) : Nothing = {"Elixir.String", "to_atom", [x]}
```

Comment:
@emilsoman I tried removing this, but to keep my Nothing test, I'll have to either introduce a `Nothing` literal or accept uppercase letters in atoms.

For now, I managed to introduce this *lovely* piece of code to force things

Reply:
This also shows that it may be better to accept `:"<content>"` syntax and also to support the same charsets as Elixir and Erlang support for atoms

Reply:
😄  haha i understand the pain here. We can support `:"string"` in the future